File: frontend/src/api/scans.ts
import apiClient from './client';
import { msalInstance } from '../auth/AuthProvider';
import { authDisabledMode } from '../auth/useAuth';
import type { ScanRun } from '../types';

export interface DelegatedTokens {
  graph?: string;
  management?: string;
  keyvault?: string;
}

/**
 * Acquire delegated tokens for Azure resources via MSAL.
 * Returns tokens for Graph, ARM, and Key Vault that the backend
 * can use to scan on behalf of the logged-in user.
 */
export async function acquireDelegatedTokens(): Promise<DelegatedTokens> {
  if (authDisabledMode || !msalInstance) {
    throw new Error('Delegated credentials require MSAL authentication');
  }

  const account = msalInstance.getActiveAccount() || msalInstance.getAllAccounts()[0];
  if (!account) throw new Error('No signed-in account');

  const acquire = async (scopes: string[]): Promise<string | undefined> => {
    try {
      const res = await msalInstance!.acquireTokenSilent({ scopes, account });
      return res.accessToken;
    } catch {
      try {
        const res = await msalInstance!.acquireTokenPopup({ scopes, account });
        return res.accessToken;
      } catch {
        return undefined;
      }
    }
  };

  // Acquire tokens sequentially (popups can't overlap)
  const graph = await acquire(['https://graph.microsoft.com/.default']);
  const management = await acquire(['https://management.azure.com/user_impersonation']);
  const keyvault = await acquire(['https://vault.azure.net/user_impersonation']);

  if (!graph && !management && !keyvault) {
    throw new Error(
      'Could not acquire any delegated tokens. Ensure your app registration ' +
      'has API permissions for Microsoft Graph, Azure Service Management, ' +
      'and/or Azure Key Vault.'
    );
  }

  return { graph, management, keyvault };
}

export async function triggerScan(useDelegated = false): Promise<{ status: string; message: string }> {
  let delegatedTokens: DelegatedTokens | undefined;

  if (useDelegated) {
    delegatedTokens = await acquireDelegatedTokens();
  }

  const { data } = await apiClient.post('/scans/trigger', {
    delegatedTokens: delegatedTokens || null,
  });
  return data;
}

export async function fetchScanHistory(page = 1, pageSize = 10): Promise<{ items: ScanRun[]; page: number; pageSize: number }> {
  const { data } = await apiClient.get('/scans/history', { params: { page, page_size: pageSize } });
  return data;
}

export async function fetchLatestScan(): Promise<ScanRun | null> {
  const { data } = await apiClient.get('/scans/latest');
  return data;
}

File: backend/app/routers/scans.py
import asyncio
import json
import logging
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, BackgroundTasks, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from app.auth.dependencies import require_admin, require_viewer
from app.db.cosmos_client import get_scan_history_container
from app.db.queries import count_items, query_items
from app.models.user import UserInfo
from app.services.scanner.event_bus import ScanEventBus

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/scans", tags=["scans"])


class DelegatedTokens(BaseModel):
    """Tokens acquired by the frontend via MSAL for user-delegated scanning."""
    graph: Optional[str] = None
    management: Optional[str] = None
    keyvault: Optional[str] = None


class ScanTriggerBody(BaseModel):
    delegatedTokens: Optional[DelegatedTokens] = None


@router.post("/trigger")
async def trigger_scan(
    background_tasks: BackgroundTasks,
    body: ScanTriggerBody = ScanTriggerBody(),
    user: UserInfo = Depends(require_admin),
):
    """Trigger an immediate full scan.

    Optionally accepts delegated tokens to scan using the logged-in user's
    permissions instead of the service principal.

    Returns the scan ID so the client can connect to the SSE stream.
    """
    from app.services.scanner.orchestrator import run_full_scan

    scan_id = str(uuid.uuid4())

    credential = None
    if body.delegatedTokens:
        from app.utils.azure_credential import DelegatedTokenCredential
        tokens = body.delegatedTokens.model_dump(exclude_none=True)
        if tokens:
            credential = DelegatedTokenCredential(tokens)
            logger.info(f"Scan triggered with delegated credentials by {user.email} "
                        f"(resources: {list(tokens.keys())})")

    # Create the event bus before starting the background task so SSE clients
    # can subscribe immediately and not miss early events.
    ScanEventBus.create(scan_id)

    background_tasks.add_task(run_full_scan, triggered_by=user.email, credential=credential, scan_id=scan_id)
    mode = "delegated" if credential else "service_principal"
    return {
        "status": "started",
        "scanId": scan_id,
        "message": f"Full scan triggered in background ({mode} credentials)",
    }


@router.get("/active")
async def get_active_scan(user: UserInfo = Depends(require_viewer)):
    """Return the currently running scan ID, if any."""
    scan_id = ScanEventBus.get_active_scan_id()
    if scan_id:
        return {"scanId": scan_id, "active": True}
    return {"scanId": None, "active": False}


@router.get("/stream/{scan_id}")
async def stream_scan_logs(scan_id: str, request: Request, user: UserInfo = Depends(require_viewer)):
    """SSE endpoint that streams real-time scan progress events.

    Replays buffered history for late-joining clients, then streams live events.
    Closes when the scan completes or the client disconnects.
    """
    bus = ScanEventBus.get(scan_id)
    if not bus:
        return StreamingResponse(
            iter([f"data: {json.dumps({'type': 'error', 'message': 'Scan not found or already cleaned up'})}\n\n"]),
            media_type="text/event-stream",
        )

    queue = bus.subscribe()

    async def event_generator():
        try:
            while True:
                if await request.is_disconnected():
                    break
                try:
                    event = await asyncio.wait_for(queue.get(), timeout=30)
                except asyncio.TimeoutError:
                    # Send keepalive comment to prevent proxy/browser timeouts
                    yield ": keepalive\n\n"
                    continue

                if event is None:
                    break  # scan completed

                yield f"data: {json.dumps(event.to_dict())}\n\n"
        finally:
            bus.unsubscribe(queue)

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


@router.get("/history")
async def get_scan_history(
    page: int = 1,
    page_size: int = 10,
    user: UserInfo = Depends(require_viewer),
):
    """List past scan runs."""
    container = get_scan_history_container()

    # Get total count
    total = await count_items(container, "SELECT VALUE COUNT(1) FROM c")

    query = "SELECT * FROM c ORDER BY c.startedAt DESC OFFSET @offset LIMIT @limit"
    params = [
        {"name": "@offset", "value": (page - 1) * page_size},
        {"name": "@limit", "value": page_size},
    ]
    items = await query_items(container, query, params)
    return {"items": items, "page": page, "pageSize": page_size, "total": total}


@router.get("/latest")
async def get_latest_scan(user: UserInfo = Depends(require_viewer)):
    """Get the most recent scan result."""
    container = get_scan_history_container()
    query = "SELECT TOP 1 * FROM c ORDER BY c.startedAt DESC"
    results = await query_items(container, query)
    return results[0] if results else None
